src/scripts/shards/Shards.ts

```diff
@@ -8,6 +8,24 @@ class Shards implements Feature {
         GameHelper.enumLength(PokemonType) - 1;
     public static readonly nEffects: number =
         GameHelper.enumLength(GameConstants.TypeEffectiveness);
+        
+    private static readonly invalidUpgrades: Map<PokemonType, Set<GameConstants.TypeEffectiveness>> =
+        
+        new Map([
+            // Immune
+            [PokemonType.Fire, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Water, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Grass, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Ice, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Flying, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Bug, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Rock, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Dark, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Steel, new Set([GameConstants.TypeEffectiveness.Immune])],
+            [PokemonType.Fairy, new Set([GameConstants.TypeEffectiveness.Immune])],
+            // Super Effective
+            [PokemonType.Normal, new Set([GameConstants.TypeEffectiveness.Very])],
+        ]);
```

Comment:
Are we able to just automatically calculate this from the type effectiveness array?

In the unlikely event a new type is added, or effectiveness changed, it will then automatically just work.